src/celcompat/cc.cpp

```diff
@@ -9,66 +9,254 @@
+        std::memcpy(buffer, ptr, length);
+        buffer += length;
+    }
+    else if (*ptr == 'n' || *ptr == 'N')
+    {
+        if (last - ptr < 3) { return { first, std::errc::invalid_argument }; }
+        std::memcpy(buffer, ptr, 3);
+        ptr += 3;
+        buffer += 3;
+        if (ptr < last && *ptr == '(')
+        {
+            *(buffer++) = *(ptr++);
+            while (ptr < last)
+            {
+                if (buffer == buffer_end)
+                {
+                    // non-standard return code: no space to put string terminator
+                    return { first, std::errc::not_enough_memory };
+                }
+                if (*ptr == ')')
+                {
+                    *(buffer++) = *(ptr++);
+                    break;
+                }
+                else if (std::isalnum(static_cast<unsigned char>(*ptr))
+                    || *ptr == '_')
+                {
+                    *(buffer++) = *(ptr++);
+                }
+                else
+                {
+                    break;
+                }
+            }
+        }
+    }
+    else
+    {
+        if (fmt == chars_format::hex)
+        {
+            hex_prefix = true;
+            *(buffer++) = '0';
+            *(buffer++) = 'x';
+        }
+
+        State state = State::Start;
+        while (ptr < last)
+        {
+            if (*ptr == '.')
+            {
+                if (state != State::Start) { break; }
+                state = State::Fraction;
+            }
+            else if (*ptr == '+' || *ptr == '-')
+            {
+                if (state != State::ExponentStart) { break; }
+                state = State::Exponent;
+            }
+            else if (fmt == chars_format::hex)
+            {
+                if (*ptr == 'p' || *ptr == 'P')
+                {
+                    if (state != State::Start && state != State::Fraction) { break; }
+                    state = State::ExponentStart;
+                }
+                else if (std::isxdigit(static_cast<unsigned char>(*ptr)))
+                {
+                    if (state == State::ExponentStart) { state = State::Exponent; }
+                }
+                else
+                {
+                    break;
+                }
+            }
+            else if (*ptr == 'e' || *ptr == 'E')
+            {
+                if ((fmt & chars_format::scientific) != chars_format::scientific) { break; }
+                state = State::ExponentStart;
+            }
+            else if (std::isdigit(static_cast<unsigned char>(*ptr)))
+            {
+                if (state == State::ExponentStart) { state = State::Exponent; }
+            }
+            else
+            {
+                break;
+            }
+
+            if (buffer == buffer_end) {
+                // non-standard return code: no space to put string terminator
+                return { first, std::errc::not_enough_memory };
```

Comment:
you reserve 1 char for '\0' on line 45, so i'm not sure that you should return error when buffer == buffer_end

Reply:
This indicates there isn't enough space in the buffer to parse the literal. While there is space to write the terminating zero byte to avoid out-of-bounds reads, it will still give the wrong result, e.g. if some value at the 1000th place would change the rounding. Hence this is an error condition.

Reply:
Added more descriptive comments.